persist/fs/commitlog/reader.go

```diff
@@ -109,58 +158,166 @@ func (r *reader) Read() (
+func (r *reader) readLoop() {
+	index := 0
+	eofFound := false
+	for {
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			r.shutdownChan <- r.close()
 			return
+		default:
+			if eofFound {
+				continue
+			}
+			data, err := r.readChunk()
+			if err == io.EOF {
+				eofFound = true
+			}
+
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+
+			index++
 		}
+	}
+}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		metadata = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       decoded.Shard,
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- *readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := &readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			outBuf <- readResponse
+			continue
+		}
+		decoder.Reset(arg.bytes.Get())
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			outBuf <- readResponse
+			continue
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
-	}
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
 
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
-	}
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadataLock.Lock()
+			r.metadataLookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       decoded.Shard,
+			}
+			waitGroup, ok := r.metadataWaitGroups[entry.Index]
+			// Another goroutine is blocked waiting for this metadata
+			if ok {
+				waitGroup.Done()
+			}
+			r.metadataLock.Unlock()
+			namespace.DecRef()
+			id.DecRef()
+		}
+
+		var pendingResult *sync.WaitGroup
+		var metadata Series
+		var hasMetadata bool
+		var ok bool
+
+		// Try cheap read lock first
+		r.metadataLock.RLock()
+		metadata, hasMetadata = r.metadataLookup[entry.Index]
+		r.metadataLock.RUnlock()
+
+		// The required metadata hasn't been processed yet
+		r.metadataLock.Lock()
```

Comment:
You can move this write lock one level deeper to avoid a write lock if the first read lock succeeds, e.g.:
```go
rwmutex.RLock()
value, ok := r.lookup[key]
rwmutex.RUnlock()

if !ok {
  rwmutex.Lock()
  value, ok = r.lookup[key]
  if !ok {
    // insert
  }
  rwmutex.Unlock()
}
```

Reply:
oh duh, thanks